libflux/src/semantic/flatbuffers/mod.rs

```diff
@@ -0,0 +1,672 @@
+//! This module defines methods for serializing and deserializing MonoTypes
+//! and PolyTypes using the flatbuffer encoding.
+//!
+mod semantic_generated;
+use semantic_generated::fbsemantic as fb;
+
+use flatbuffers;
+use std::collections::HashMap;
+
+#[rustfmt::skip]
+use crate::semantic::types::{
+    Array,
+    Function,
+    Kind,
+    MonoType,
+    PolyType,
+    Property,
+    Row,
+    Tvar,
+};
+
+/// Decodes a PolyType from a flatbuffer
+impl From<fb::PolyType<'_>> for PolyType {
+    fn from(t: fb::PolyType<'_>) -> PolyType {
+        let v = t.vars().unwrap();
+        let mut vars = Vec::new();
+        for i in 0..v.len() {
+            vars.push(v.get(i).into());
+        }
+        let c = t.cons().unwrap();
+        let mut cons = HashMap::new();
+        for i in 0..c.len() {
+            let (tv, kind) = c.get(i).into();
+            cons.entry(tv).or_insert(Vec::new()).push(kind);
+        }
+        PolyType {
+            vars,
+            cons,
+            expr: from_table(t.expr().unwrap(), t.expr_type()),
+        }
+    }
+}
+
+impl From<fb::Constraint<'_>> for (Tvar, Kind) {
+    fn from(constraint: fb::Constraint) -> (Tvar, Kind) {
+        (constraint.tvar().unwrap().into(), constraint.kind().into())
+    }
+}
+
+impl From<fb::Kind> for Kind {
+    fn from(kind: fb::Kind) -> Kind {
+        match kind {
+            fb::Kind::Addable => Kind::Addable,
+            fb::Kind::Subtractable => Kind::Subtractable,
+            fb::Kind::Divisible => Kind::Divisible,
+            fb::Kind::Comparable => Kind::Comparable,
+            fb::Kind::Equatable => Kind::Equatable,
+            fb::Kind::Nullable => Kind::Nullable,
+        }
+    }
+}
+
+impl From<Kind> for fb::Kind {
+    fn from(kind: Kind) -> fb::Kind {
+        match kind {
+            Kind::Addable => fb::Kind::Addable,
+            Kind::Subtractable => fb::Kind::Subtractable,
+            Kind::Divisible => fb::Kind::Divisible,
+            Kind::Comparable => fb::Kind::Comparable,
+            Kind::Equatable => fb::Kind::Equatable,
+            Kind::Nullable => fb::Kind::Nullable,
+        }
+    }
+}
+
+fn from_table(table: flatbuffers::Table, t: fb::MonoType) -> MonoType {
+    match t {
+        fb::MonoType::Var => MonoType::Var(fb::Var::init_from_table(table).into()),
+        fb::MonoType::Arr => MonoType::Arr(Box::new(fb::Arr::init_from_table(table).into())),
+        fb::MonoType::Row => MonoType::from(fb::Row::init_from_table(table)),
+        fb::MonoType::Fun => MonoType::Fun(Box::new(fb::Fun::init_from_table(table).into())),
+        fb::MonoType::Basic => MonoType::from(fb::Basic::init_from_table(table)),
+        fb::MonoType::NONE => panic!("expected MonoType encoding; got NONE"),
```

Comment:
I do not. I believe it is undefined since a panic in Rust when called from C is undefined. At least that's what I have read. The reason I kept this panic here is that I assumed it was essentially dead code as this arm will never be reached if the flatbuffer is constructed using the serialization code.

But perhaps I should remove it just to be totally sure undefined behavior is not possible.